### PR TITLE
retain OpTreeContext structure for consistency with scala-2 

### PR DESCRIPTION
**Pull request by jrudolph**
And also because it simplifies the AST construction.

WDYT, @yanns? Would be good to merge this first otherwise, we'll get lots of conflicts.

### CODE REVIEW

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 1]
```diff
@@ -42,85 +42,39 @@ private[parboiled2] trait RuleRunnable {
...
-
-  private def reportError(error: String, expr: Expr[Any])(using quotes: Quotes): Nothing = {
-    quotes.reflect.report.error(error, expr)
-    throw new Exception(error)
-  }
-
+import scala.quoted._
+class OpTreeContext(parser: Expr[Parser])(using Quotes) {
   sealed trait OpTree {
-    def parser: Expr[Parser]
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:92]: We don't need to capture this now anymore, it's captured once per macro and then available from the context.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 2]
```diff
@@ -42,85 +42,39 @@ private[parboiled2] trait RuleRunnable {
...
-    def render(wrapped: Boolean)(using Quotes): Expr[Boolean]
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:93]: Same with `using Quotes` which can be provided from the context.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 3]
```diff
@@ -18,17 +18,17 @@ package org.parboiled2
 
 import org.parboiled2.support.hlist.HList
 
-private[parboiled2] trait ParserMacroMethods {
+private[parboiled2] trait ParserMacroMethods { parser: Parser =>
```
Comment (yanns) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:21]: clever! 👍

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 4]
```diff
@@ -42,85 +42,39 @@ private[parboiled2] trait RuleRunnable {
...
-    }
-  }
-
-  private def reportError(error: String, expr: Expr[Any])(using quotes: Quotes): Nothing = {
-    quotes.reflect.report.error(error, expr)
-    throw new Exception(error)
-  }
-
+import scala.quoted._
+class OpTreeContext(parser: Expr[Parser])(using Quotes) {
```
Comment (yanns) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:46]: - In scala-2, `OpTreeContext` is defined in its own file under `org.parboiled2.support`. Should we keep the same structure?
- I though having read some best practices about avoiding capturing the `Quotes` in a class with scala 3 macros, but I could not find this anymore. Maybe it was https://docs.scala-lang.org/scala3/guides/macros/best-practices.html, but was removed? Anyway, I'm good with that.

**Comment by jrudolph** [comment]
(Sorry for the ping/request for review, @xuwei-k, that was an accident)

**Comment by jrudolph** [comment]
Do you have any ongoing work that would conflict with this, @yanns? Otherwise, I'd want to soon go ahead and merge it.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala (hunk 4 above)
Comment (jrudolph) [reply parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:46]: > In scala-2, `OpTreeContext` is defined in its own file under `org.parboiled2.support`. Should we keep the same structure?

Ultimately, yes. But I like being able to move around quickly right now. We can extract it later.

>     * I though having read some best practices about avoiding capturing the `Quotes` in a class with scala 3 macros, but I could not find this anymore. Maybe it was [docs.scala-lang.org/scala3/guides/macros/best-practices.html](https://docs.scala-lang.org/scala3/guides/macros/best-practices.html), but was removed? Anyway, I'm good with that.

Didn't find anything about that but I guess we'll find out if that should be a problem

**Pull request merged**